File: worker/src/settle.ts
/** Feed-driven settlement. Port of bot/settle.py — never guesses. */

export interface FinalScore {
  home: number;
  away: number;
}

export interface SettleSelection {
  market_type: string;
  outcome: string;
  line: string | null;
}

export function settle1x2(selection: SettleSelection, final: FinalScore | null): string {
  if (!final || final.home === undefined || final.away === undefined) return "pending";
  const { home, away } = final;
  const outcome = (selection.outcome || "").toLowerCase();
  const winner = home === away ? "draw" : home > away ? "home" : "away";
  if (outcome === "yes" || outcome === "no") return "pending"; // prediction-market semantics need market-specific logic
  return outcome === winner ? "won" : "lost";
}

export function settleTotal(selection: SettleSelection, final: FinalScore | null): string {
  if (!final || selection.line === null || selection.line === undefined) return "pending";
  const total = final.home + final.away;
  const line = Number(selection.line);
  if (!Number.isFinite(total) || !Number.isFinite(line)) return "pending";
  const outcome = (selection.outcome || "").toLowerCase();
  if (total === line) return "void";
  const isOver = total > line;
  if (outcome.startsWith("over") || outcome === "yes") return isOver ? "won" : "lost";
  if (outcome.startsWith("under") || outcome === "no") return !isOver ? "won" : "lost";
  return "pending";
}

export function settle(selection: SettleSelection, final: FinalScore | null): string {
  const mt = (selection.market_type || "").toLowerCase();
  if (mt === "moneyline" || mt === "1x2") return settle1x2(selection, final);
  if (mt.startsWith("total")) return settleTotal(selection, final);
  return "pending"; // exotic markets: explicit logic per type, never guessed
}

File: bot/settle.py
"""Feed-driven settlement. Unsettlable markets stay pending, never guessed."""
from __future__ import annotations

from .feeds.base import LiveEvent


def settle_1x2(selection: dict, final: dict | None) -> str:
    """Settle moneyline/1x2 given {'home': goals, 'away': goals} and outcome side."""
    if not final or "home" not in final or "away" not in final:
        return "pending"
    home, away = final["home"], final["away"]
    outcome = (selection.get("outcome") or "").lower()
    if home == away:
        winner = "draw"
    else:
        winner = "home" if home > away else "away"
    if outcome in ("yes", "no"):
        return "pending"  # prediction-market semantics need market-specific logic
    return "won" if outcome == winner else "lost"


def settle_total(selection: dict, final: dict | None) -> str:
    if not final or selection.get("line") is None:
        return "pending"
    try:
        total = final["home"] + final["away"]
        line = float(selection["line"])
    except (KeyError, TypeError, ValueError):
        return "pending"
    outcome = (selection.get("outcome") or "").lower()
    if total == line:
        return "void"
    is_over = total > line
    if outcome.startswith("over") or outcome == "yes":
        return "won" if is_over else "lost"
    if outcome.startswith("under") or outcome == "no":
        return "won" if not is_over else "lost"
    return "pending"


def settle(selection: dict, final: dict | None) -> str:
    mt = (selection.get("market_type") or "").lower()
    if mt in ("moneyline", "1x2"):
        return settle_1x2(selection, final)
    if mt.startswith("total"):
        return settle_total(selection, final)
    return "pending"  # exotic markets: explicit logic per type, never guessed
